File: 0x13-javascript_objects_scopes_closures/2-rectangle.js
#!/usr/bin/node
/* Write an empty class Rectangle that defines a rectangle
   The constructor must take 2 arguments w and h
   Initialize the instance attribute width with the value of w
   Initialize the instance attribute height with the value of h
   If w or h is equal to 0 or not a positive integer, create an empty object
*/
// set the exports object to a function or a new object
module.exports = class Rectangle {
  constructor (w, h) {
    if ((w <= 0 || w === undefined) || (h <= 0 || h === undefined)) {
      return this;
    }
    this.width = w;
    this.height = h;
  }
};

File: 0x08-python-more_classes/2-rectangle.py
#!/usr/bin/python3
"""
class Rectangle that defines a rectangle
def __init__(self, width=0, height=0):
def width(self)
width must be an integer, otherwise raise a TypeError
width is less than 0, raise a ValueError
def height(self)
height must be an integer, otherwise raise a TypeError
height is less than 0, raise a ValueError
def area(self): that returns the rectangle area
def perimeter(self):
if width or height is equal to 0, perimeter is equal to 0
"""


class Rectangle:
    """
    initialize in constructor
    """
    def __init__(self, width=0, height=0):
        self.__width = width
        self.__height = height

    @property
    def width(self):
        """  getter """
        return self.__width

    @width.setter
    def width(self, value):
        """ setter """
        if type(value) is not int:
            raise TypeError("width must be an integer")
        elif value < 0:
            raise ValueError("width must be >= 0")
        else:
            self.__width = value

    @property
    def height(self):
        """  getter """
        return self.__height

    @height.setter
    def height(self, value):
        """  setter """
        if type(value) is not int:
            raise TypeError("height must be an integer")
        elif value < 0:
            raise ValueError("height must be >= 0")
        else:
            self.__height = value

    def area(self):
        """ returns the rectangle area  """
        return self.__width * self.__height

    def perimeter(self):
        """ returns the rectangle perimeter """
        if self.__height == 0 or self.__width == 0:
            return 0
        return ((self.__width + self.__height) * 2)
